crates/circuit/src/parameterexpression.rs

```diff
@@ -0,0 +1,779 @@
+        }
+    }
+    pub fn cos(&self) -> Self {
+        Self {
+            expr: self.expr.cos(),
+        }
+    }
+    pub fn tan(&self) -> Self {
+        Self {
+            expr: self.expr.tan(),
+        }
+    }
+    pub fn asin(&self) -> Self {
+        Self {
+            expr: self.expr.asin(),
+        }
+    }
+    pub fn acos(&self) -> Self {
+        Self {
+            expr: self.expr.acos(),
+        }
+    }
+    pub fn atan(&self) -> Self {
+        Self {
+            expr: self.expr.atan(),
+        }
+    }
+    pub fn exp(&self) -> Self {
+        Self {
+            expr: self.expr.exp(),
+        }
+    }
+    pub fn log(&self) -> Self {
+        Self {
+            expr: self.expr.log(),
+        }
+    }
+    pub fn abs(&self) -> Self {
+        Self {
+            expr: self.expr.abs(),
+        }
+    }
+    pub fn sign(&self) -> Self {
+        Self {
+            expr: self.expr.sign(),
+        }
+    }
+
+    /// return value if expression does not contain any symbols
+    pub fn value(&self, py: Python) -> PyResult<PyObject> {
+        match self.expr.eval(true) {
+            Some(v) => match v {
+                Value::Real(r) => r.into_py_any(py),
+                Value::Int(i) => i.into_py_any(py),
+                Value::Complex(c) => {
+                    if (-SYMEXPR_EPSILON..SYMEXPR_EPSILON).contains(&c.im) {
+                        c.re.into_py_any(py)
+                    } else {
+                        c.into_py_any(py)
+                    }
+                }
+            },
+            None => Err(pyo3::exceptions::PyRuntimeError::new_err(
+                "Expression has some undefined symbols.",
+            )),
+        }
+    }
+
+    /// clone expression
+    pub fn copy(&self) -> Self {
+        Self {
+            expr: self.expr.clone(),
+        }
+    }
+    /// return conjugate of expression
+    pub fn conjugate(&self) -> Self {
+        Self {
+            expr: self.expr.conjugate(),
+        }
+    }
+    /// return derivative of this expression for param
+    pub fn derivative(&self, param: &Self) -> Self {
+        Self {
+            expr: self.expr.derivative(&param.expr),
+        }
+    }
+
+    /// expand expression
+    pub fn expand(&self) -> Self {
+        Self {
+            expr: self.expr.expand(),
+        }
+    }
+
+    /// get hashset of all the symbols used in this expression
+    pub fn symbols(&self) -> HashSet<String> {
+        self.expr.symbols()
+    }
+
+    /// return all values in this equation
+    pub fn values(&self, py: Python) -> PyResult<Vec<PyObject>> {
+        let ret: Vec<PyObject> = self
+            .expr
+            .values()
+            .iter()
+            .map(|val| match val {
+                Value::Real(r) => r.into_py_any(py).unwrap(),
+                Value::Int(i) => i.into_py_any(py).unwrap(),
+                Value::Complex(c) => c.into_py_any(py).unwrap(),
+            })
+            .collect();
+        Ok(ret)
```

Comment:
You should be able to propagate the error like this since you can collect an iterator of `Result` into a `Result<Vec<_>>`.

```suggestion
        self
            .expr
            .values()
            .iter()
            .map(|val| match val {
                Value::Real(r) => r.into_py_any(py),
                Value::Int(i) => i.into_py_any(py),
                Value::Complex(c) => c.into_py_any(py),
            })
            .collect()
```